### ISSUE TITLE
Crash in Mesh Warp

### ISSUE THREAD

**Issue by akenmorris** [SCIInstitute/ShapeWorks#964]
See discussion here: #898 

```
__pthread_kill 0x00007fff6880333a
pthread_kill 0x00007fff688bfe60
abort 0x00007fff6878a808
__assert_rtn 0x00007fff68789ac6
Eigen::DenseCoeffsBase<Eigen::Matrix<int, -1, 1, 0, -1, 1>, 1>::operator()(long) DenseCoeffsBase.h:425
bool igl::min_quad_with_fixed_precompute<double, Eigen::Matrix<int, -1, 1, 0, -1, 1> >(Eigen::SparseMatrix<double, 0, int> const&, Eigen::MatrixBase<Eigen::Matrix<int, -1, 1, 0, -1, 1> > const&, Eigen::SparseMatrix<double, 0, int> const&, bool, igl::min_quad_with_fixed_data<double>&) min_quad_with_fixed.cpp:83
bool igl::min_quad_with_fixed<double, Eigen::Matrix<int, -1, 1, 0, -1, 1>, Eigen::Matrix<double, -1, 1, 0, -1, 1>, Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<double, -1, 1, 0, -1, 1>, Eigen::Matrix<double, -1, -1, 0, -1, -1> >(Eigen::SparseMatrix<double, 0, int> const&, Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0, -1, 1> > const&, Eigen::MatrixBase<Eigen::Matrix<int, -1, 1, 0, -1, 1> > const&, Eigen::MatrixBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> > const&, Eigen::SparseMatrix<double, 0, int> const&, Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0, -1, 1> > const&, bool, Eigen::PlainObjectBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> >&) min_quad_with_fixed.cpp:578
bool igl::biharmonic_coordinates<Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<int, -1, -1, 0, -1, -1>, int, Eigen::Matrix<double, -1, -1, 0, -1, -1> >(Eigen::MatrixBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> > const&, Eigen::MatrixBase<Eigen::Matrix<int, -1, -1, 0, -1, -1> > const&, std::__1::vector<std::__1::vector<int, std::__1::allocator<int> >, std::__1::allocator<std::__1::vector<int, std::__1::allocator<int> > > > const&, int, Eigen::PlainObjectBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> >&) biharmonic_coordinates.cpp:200
W_precomputation(Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<int, -1, -1, 0, -1, -1>, Eigen::Matrix<int, -1, -1, 0, -1, -1>) WarpMeshUsingLandmarks.cxx:56
main WarpMeshUsingLandmarks.cxx:184
start 0x00007fff686bbcc9
```

@riddhishb wrote:

I think I have figured out the issue, the algorithm is interpolating the value of particle to the closest mesh vertex first and then using TPS to interpolate to other vertices, the for loop contining "unknown vertices" do not have a added contion as to what happens when two different control points gets projected to the same nearest vertex.
This as far as I can think can happen in two senarios,

1. one where two control points are very close together (dunno how to decide threshold maybe some percentage of mean inter point distance).

2. When the mesh is sparse and number of points is decently close to number of template mesh vertices (this happens in the data you sent Alan), an easy fix to this is just subdivide the template mesh and increase the number of vertices and that fixed the problem.

There is a way to produce a flag from inside the IGL which checks fro this phenomena, but I do not know how to change teh libigl everytime after build depndency clones a libigl.

For reference, below is a studio project exhibiting this crash.  Simply open and click "Analysis"

[warp_mesh_crash.zip](https://github.com/SCIInstitute/ShapeWorks/files/6012843/warp_mesh_crash.zip)

**Comment by riddhishb** [comment]
@akenmorris I guess we can do the projection, I will have to figure out how to perform this. I have managed to do this inside the libigl function where I essentially do the ame this by randomly keep just one of the redundant projection, it is just a single code that needs to be updated. Wouldnt that be easier?
Yeah you are right subdividing will not gaurantee this

**Comment by akenmorris** [comment]
Yeah, if it's easily to fix in LibIGL, then let's do it there.  Do we need to make a fork of the repo and target that instead?

**Comment by akenmorris** [comment]
> Do you think it is worth submitting a bug report to libigl, and get a PR out there?

Yes

**Comment by akenmorris** [comment]
@riddhishb , what's the change to LibIGL?  Can you send me the file?

**Comment by sheryjoe** [comment]
@riddhishb where are we on this one? -- needed for the release.

**Comment by riddhishb** [comment]
Hi All

I am putting here the location and fix of the bug.

The bug is in the file `min_quad_with_fixed.cpp` that can be found in `dependencies/install/libigl/include` , if we replace the lines 71-86 by the following modification the crash goes away as it selects only one of the redundant vertices and hence index overflow is prevented.

```
std::vector<bool> unknown_mask;
  unknown_mask.resize(n,true);
  int krt = 0;
  for(int i = 0;i<kr;i++)
  {
    if(unknown_mask[known(i, 0)]){krt++;}
    unknown_mask[known(i, 0)] = false;
  }
  int u = 0;
  data.unknown.resize(n-krt);
  for(int i = 0;i<n;i++)
  { 
    if(unknown_mask[i])
    { 
      data.unknown(u) = i;
      u++;
    }
  }
```

Now the problem is I do not know how to change this file? Do we add in our own file for the affected cpp file or we change the lines after the dependencies are installed.

I will be traveling from Saturday to Monday (I will be more or less available for the next 48 hours) so I might not be able to address this @akenmorris if this is holding up the release can you please address this, I am guessing this will be much quicker for you.

Thanks

**Comment by riddhishb** [comment]
There is one issue if we just select one of these vertices, it introduces some artifacts in the reconstruction so it is worthwhile to explore better fixes in the future releases and/or add a subdivision option on the template mesh.

**Comment by akenmorris** [comment]
Thanks @riddhishb , I will take care of it.  We'll focus on the crash for now and add a separate issue for the reconstruction artifact.

**Comment by riddhishb** [comment]
Thanks a ton @akenmorris

**Comment by akenmorris** [comment]
Resolved in #1035
